python/pyarrow/scalar.pxi

```diff
@@ -731,6 +731,57 @@ cdef class LargeListValue(ArrayValue):
         return result
 
 
+cdef class MapValue(ArrayValue):
+    """
+    Concrete class for map array elements.
+    """
+
+    def __len__(self):
+        """
+        Return the number of values.
+        """
+        return self.length()
+
+    def __getitem__(self, i):
+        """
+        Return the value at the given index.
+        """
+        return self.getitem(_normalize_index(i, self.length()))
+
+    def __iter__(self):
+        """
+        Iterate over this element's values.
+        """
+        for i in range(len(self)):
+            yield self.getitem(i)
+        raise StopIteration
+
+    cdef void _set_array(self, const shared_ptr[CArray]& sp_array):
+        self.sp_array = sp_array
+        self.ap = <CMapArray*> sp_array.get()
+        self.value_type = pyarrow_wrap_data_type(self.ap.value_type())
+
+    cdef getitem(self, int64_t i):
+        cdef int64_t j = self.ap.value_offset(self.index) + i
+        return box_scalar(self.value_type, self.ap.values(), j)
+
+    cdef int64_t length(self):
+        return self.ap.value_length(self.index)
+
+    def as_py(self):
+        """
+        Return this value as a Python list of dicts with pairs.
```

Comment:
This could also be a single dict with multiple key-item pairs ? 

(I am not familiar enough with a "map type" to know if that is a more natural representation, though)

Reply:
Ah, I see below in the tests this is actually a list like `[{'key': key1, 'value': value1}, {'key': key2, 'value': value2}, ..]` and not `[{key1: value1}, {key2: value2}, ...]` (as I had in my head)

So with such a structure, it of course needs to be a list as it has duplicate dict keys. But still wondering if something like `{key1: value1, key2, value2}` is not more natural for Python?

Reply:
One trouble with this is that in Schema.fbs we say "the application is responsible for ensuring that the keys are hashable and unique". What about returning a list of 2-tuples?

Reply:
Maybe there could be two methods `as_dict` and `as_list`.

`as_list` would return a list of key/value dicts:
```
[{'key': key1, 'value': value1}, {'key': key2, 'value': value2}, ..]
```

whereas `to_dict` would return a single python dictionary:
```
{'key1': value1, 'key2': value2, ...}
```

The `to_dict` method could then raise an exception if the keys weren't hashable or unique.

Reply:
> What about returning a list of 2-tuples?

We could do that, but I thought it would be best to return the same as a list of structs for the default.